pczt/src/transparent.rs

```diff
@@ -0,0 +1,388 @@
+    pub(crate) inputs: Vec<Input>,
+    pub(crate) outputs: Vec<Output>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Input {
+    //
+    // Transparent effecting data.
+    //
+    // These are required fields that are part of the final transaction, and are filled in
+    // by the Constructor when adding an output.
+    //
+    pub(crate) prevout_txid: [u8; 32],
+    pub(crate) prevout_index: u32,
+
+    /// The sequence number of this input.
+    ///
+    /// - This is set by the Constructor.
+    /// - If omitted, the sequence number is assumed to be the final sequence number
+    ///   (`0xffffffff`).
+    pub(crate) sequence: Option<u32>,
+
+    /// The minimum Unix timstamp that this input requires to be set as the transaction's
+    /// lock time.
+    ///
+    /// - This is set by the Constructor.
+    /// - This must be greater than or equal to 500000000.
+    pub(crate) required_time_lock_time: Option<u32>,
+
+    /// The minimum block height that this input requires to be set as the transaction's
+    /// lock time.
+    ///
+    /// - This is set by the Constructor.
+    /// - This must be greater than 0 and less than 500000000.
+    pub(crate) required_height_lock_time: Option<u32>,
+
+    /// A satisfying witness for the `script_pubkey` of the input being spent.
+    ///
+    /// This is set by the Spend Finalizer.
+    pub(crate) script_sig: Option<Vec<u8>>,
+
+    // These are required by the Transaction Extractor, to derive the shielded sighash
+    // needed for computing the binding signatures.
+    pub(crate) value: u64,
+    pub(crate) script_pubkey: Vec<u8>,
+
+    /// The script required to spend this output, if it is P2SH.
+    ///
+    /// Set to `None` if this is a P2PKH output.
+    pub(crate) redeem_script: Option<Vec<u8>>,
+
+    /// A map from a pubkey to a signature created by it.
+    ///
+    /// - Each pubkey should appear in `script_pubkey` or `redeem_script`.
+    /// - Each entry is set by a Signer, and should contain an ECDSA signature that is
+    ///   valid under the corresponding pubkey.
+    /// - These are required by the Spend Finalizer to assemble `script_sig`.
+    pub(crate) partial_signatures: BTreeMap<[u8; 33], Vec<u8>>,
+
+    /// The sighash type to be used for this input.
+    ///
+    /// - Signers must use this sighash type to produce their signatures. Signers that
+    ///   cannot produce signatures for this sighash type must not provide a signature.
+    /// - Spend Finalizers must fail to finalize inputs which have signatures that do not
+    ///   match this sighash type.
+    pub(crate) sighash_type: u8,
+
+    /// A map from a pubkey to the BIP 32 derivation path at which its corresponding
+    /// spending key can be found.
+    ///
+    /// - The pubkeys should appear in `script_pubkey` or `redeem_script`.
+    /// - Each entry is set by an Updater.
+    /// - Individual entries may be required by a Signer.
+    pub(crate) bip32_derivation: BTreeMap<[u8; 33], Zip32Derivation>,
```

Comment:
It's a little bit odd to me that this uses ZIP 32 derivation; for something like signing transactions that spend outputs to legacy `zcashd` wallet UTXOs where the spending key was drawn from randomness rather than ZIP 32 derivation, would the process be to just look up from the pubkey and ignore the value in the map?

Reply:
In that case, the map would be empty. There is no need to put anything in the map; it's purely to help the Signer figure out where the key material is, but if they already know then there's no need.

Reply:
I suggested a comment to that effect.